### DISCUSSION TITLE
Feature request: Option to combine stats for pages with and without trailing slashes

### DISCUSSION THREAD

**Discussion by theodorechu** [plausible/analytics#336] [Anything Else]
## Precheck

Please note that this tracker is only for feature requests. Do not use the issue tracker for help or support. [Our docs](https://docs.plausible.io/) are a great place for most answers, but if you can’t find your answer there, you can [contact us](https://plausible.io/contact). Have an issue with your self-hosted install? You can ask in [our community forum](https://plausible.discourse.group/). Thanks!

## Prerequisites
- [x] I have searched open and closed issues to make sure that the feature has not yet been requested.

## Feature request
**Is your feature request related to a problem? Please describe.**
My website does not add trailing slashes to the end of each URL (e.g., example.com/hello vs example.com/hello/). Plausible counts them separately. For example, see https://plausible.io/docs.plausible.io/pages

**How would you like it to work**
I would like an option to combine the stats for pages with and without trailing slashes

**Screenshots (If applicable)**
If applicable, add screenshots to help explain the proposed solution.

**Comment by ukutaht** [comment]
Thanks for the suggestion, I've noticed this myself on https://plausible.io/docs.plausible.io/pages

If possible, I think the best solution is to change the website itself to redirect to ensure a consistent URL style.

If not then I agree, there should be a way to merge analytics for paths with trailing slash and ones without.

**Comment by svprdga** [comment]
The problem is not with the website itself, the problem is with third party services that manipulate the URLs.

For instance, my website has all the endpoints without the final slash, but I submit some URLs of my site to third party services that alter them, and then serve the same URL with a final slash.

That's why I think that's very critical to add an option to merge this kind of URLs, to have reliable stats.

**Comment by davidsteeb** [comment]
Is there an option to set the "pageview url" that's being tracked? That could enable developers to set the url for pageviews to the canonical for example.

**Comment by ukutaht** [comment]
@davidsteeb Not at the moment. Makes sense to add that option though

**Comment by mikekreeki** [comment]
Having the same issue. Many frameworks on the frontend produce URLs without trailing slashes when doing history transitions, whereas if someone visits the URL directly the backend tooling (whether it is Netlify router or Heroku static buildpack) adds the slash. 

Would be super cool to have some toggle in the settings whether to merge these or not. +1 to that!

**Reply by ukutaht** [reply]
Would it be too aggressive to strip trailing slash by default and instead provide an option to preserve it? I might be projecting my own preferences to others but I feel like most people want clean URLs(without trailing slashes) for both their site and analytics.

**Reply by ukutaht** [reply]
We should always provide an option, the question is what the default should be.

**Reply by mikekreeki** [reply]
@ukutaht Sure, would work for me the best. I was not meaning to suggest a solution, just wanted to give +1 to this issue. Thanks for looking into it, any solution would be great!

**Comment by metmarkosaric** [comment]
At least from the SEO perspective, this should be done on the site itself as having two versions of the same page (http or https, www or non-www, / or non-/) can be treated by Google as two different pages which can cause the duplicate content penalty from the search algorithm. 

Something like this in the .htacess file fixes the trailing slash (most CMS have this sorted out of the box too):

RewriteCond %{REQUEST_FILENAME} !-d 
RewriteRule ^(.*)/$ /$1 [L,R=301]

We could fix this automatically in our reports (combine them both into one) but the issue still remains for the site which should be fixed.

(more info: https://ahrefs.com/blog/trailing-slash/)

**Reply by ukutaht** [reply]
From this perspective rolling them up by default would rob the user of the opportunity to find and fix trailing slash issues on their site. Would be really cool to even flag it to the user when we see two equivalent URLs with just a trailing slash difference.

And if there is an option to combine trailing slash stats it should be an explicit opt-in.

**Reply by metmarkosaric** [reply]
Would definitely be cool from seo perspective. Companies pay a lot of money for technical seo audits and this would be one of the things that is flagged in the audits.

**Reply by Vigasaurus** [reply]
This could be a cool thing to include with all the Web Vitals stuff - i.e. also include SEO/similar suggestions as recommended by any tool that we did Web Vitals with (and include this and other similar if we have data showing it).

**Reply by mffap** [reply]
IMO you should fix the issue of the user: seeing the same "content" as separate entries. Flagging a potential concern with SEO would be nice (but SEO should not be priority of the tool). We have the same issue, I just want the tracking to show me stats for the same page (in terms of "content", abstact from the technical url slug)

**Reply by sahandnayebaziz** [reply]
I'm in the same boat as @mffap, where in my use case this is frustrating as a counting issue, not really as an SEO-validity issue. If this is purely an SEO problem that I should be fixing with redirects (or another fix), maybe Plausible could link to a Google-approved or other authoritative guide on how to fix the problem

**Reply by metmarkosaric** [reply]
best places for this type of advice are seo tools such as ahrefs: https://ahrefs.com/blog/301-redirects/. here's a bit more info from google on this as well that ahrefs linked to: https://developers.google.com/search/docs/advanced/crawling/consolidate-duplicate-urls

**Comment by JMTK** [comment]
Does anyone have a 1-off script to update the existing records that have the trailing slash after I added the Rewrite rule to the htaccess? I don't know enough about the Plausible setup to do it. I tried searching postgres tables but it seems like it's not stored there

**Reply by ukutaht** [reply]
I don't have a script but some guidance if you want to figure it out.

Stats are stored in Clickhouse. [See here](https://github.com/plausible/analytics/discussions/621#discussioncomment-293169) how to connect to your clickhouse instance. It talks SQL so should be familiar.

The columns that store page urls are `events->pathname`, `sessions->entry_page`  and `sessions->exit_page`.

Clickhouse does not support regular `UPDATE` statements. Mutating data works asynchronously using the `ALTER TABLE UPDATE` command: https://clickhouse.tech/docs/en/sql-reference/statements/alter/update/

**Comment by MarcoBuster** [comment]
Any updates / ETA / suggestions on this?

**Reply by metmarkosaric** [reply]
hi @MarcoBuster! sorry, no updates/ETA. the suggestion is still the same as described above https://github.com/plausible/analytics/discussions/336#discussioncomment-400333

**Reply by MarcoBuster** [reply]
@metmarkosaric thank you :)

**Comment by sebastianbergmann** [comment]
I started to use Plausible today for `https://phpunit.de/`, the website for the PHPUnit Open Source project.

The only thing that nags me so far is the fact that Plausible treats `https://phpunit.de/` and `https://phpunit.de/index.html` as separate pages. The same HTML document is served for both URLs. This HTML document has `<link rel="canonical" href="https://phpunit.de/index.html" />` in its `<head>` section and it would be nice if Plausible would consider this information and treat `https://phpunit.de/` and `https://phpunit.de/index.html` as the same page.

To work around this issue on my end, I can either configure an HTTP redirect from `https://phpunit.de/` to `https://phpunit.de/index.html` or use `https://phpunit.de/` instead of `https://phpunit.de/index.html` for the internal links. Both solutions somehow feel wrong to me.

**Comment by chrism2671** [comment]
I would go out on a limb here and suggest this shouldn't be fixed. 

The reason is that the slash and non-slash versions are different URLs with respect to HTTP standards and, as a result, should be recorded separately. 

So, it would be better for plausible to be able to merge these results in the view layer.
